src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -297,35 +332,206 @@ private BoundExpression RewriteStringConcatenationOneExpr(BoundExpression lowere
             }
         }
 
-        private BoundExpression RewriteStringConcatenationTwoExprs(SyntaxNode syntax, BoundExpression loweredLeft, BoundExpression loweredRight)
+        private BoundExpression RewriteStringConcatenationTwoExprs(SyntaxNode syntax, BoundExpression loweredLeft, BoundExpression loweredRight, ArrayBuilder<LocalSymbol> previousLocals)
         {
-            Debug.Assert(loweredLeft.HasAnyErrors || loweredLeft.Type is { } && loweredLeft.Type.IsStringType());
-            Debug.Assert(loweredRight.HasAnyErrors || loweredRight.Type is { } && loweredRight.Type.IsStringType());
+            var leftType = loweredLeft.Type;
+            var rightType = loweredRight.Type;
+
+            Debug.Assert(loweredLeft.HasAnyErrors || leftType is not null && (leftType.IsStringType() || leftType.IsCharType()));
+            Debug.Assert(loweredRight.HasAnyErrors || rightType is not null && (rightType.IsStringType() || rightType.IsCharType()));
+
+            var leftIsChar = leftType?.IsCharType() == true;
+            var rightIsChar = rightType?.IsCharType() == true;
+
+            if ((leftIsChar || rightIsChar) &&
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_String__Concat_ReadOnlySpanReadOnlySpan, out MethodSymbol stringConcatSpans, isOptional: true) &&
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_String__op_Implicit_ToReadOnlySpanOfChar, out MethodSymbol stringImplicitConversionToReadOnlySpan, isOptional: true) &&
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_ReadOnlySpan_T__ctor_Reference, out MethodSymbol readOnlySpanCtorRefParamGeneric, isOptional: true))
+            {
+                // One of args is a char. We can use span-based concatenation, which takes a bit more IL, but avoids allocation of a string from ToString call.
+                // And since implicit conversion from string to span is a JIT intrinsic, the resulting code ends up being faster than the one generated from the "naive" case with ToString call
+                var readOnlySpanOfChar = readOnlySpanCtorRefParamGeneric.ContainingType.Construct(_compilation.GetSpecialType(SpecialType.System_Char));
+                var readOnlySpanCtorRefParamChar = readOnlySpanCtorRefParamGeneric.AsMember(readOnlySpanOfChar);
+
+                var locals = ArrayBuilder<LocalSymbol>.GetInstance(capacity: 2);
```

Comment:
Hardcoded capacity is intentional here. We are in "concat 2 expressions" method thus it can have at max 2 additional locals (for the first and the second argument). On the calee side we `Debug.Assert` that `previousLocals` contains at max 1 local from the previous iteration (though I don't think we ever have a previous local in "concat 2 expressions", but when there are 3 or 4 expressions, we can definitely have locals from previous iterations). Maybe we should move that `Debug.Assert` here to make it clearer?

Reply:
> On the calee side we Debug.Assert that previousLocals contains at max 1 local from the previous iteration

Well, that, plus 2 new locals here, means a total of 3. Let's just add the count from `previousLocals` uniformly.

Reply:
Well, `previousLocals` contain locals from previous iterations. Take `s + c1 + c2 + c3` as an example, where all `c`s require a local when constructing a span from them:
- On the first iteration we lower `s + c1`. First, there can be no previous locals here since this is the first iteration. Second, only 1 new local can appear when lowering the expression since in order to be a string concat at least 1 operand must be a string (if there are 2 char operands (`c1 + c2`) both operands are implicitly converted to `int`s and `+` operator works as number addition). As a result of this iteration we get sequence `{ locals: temp1; result: string.Concat(s, { sideEffects: temp1 = c1; result: new ReadOnlySpan<char>(in temp1) })}`
- On the second iteration we need to lower concatenation of the previous result and remaining operand `s`. We unwrap outer sequence from a previous iteration back and get `temp1` as a previous local and `s` and `{ sideEffects: temp1 = c1; result: new ReadOnlySpan<char>(in temp1) }` as previous arguments. Now we need to lower it to `string.Concat` of 3 arguments. Note, that only 1 additional local can appear here if the third argument is a char. Since in our example it is a char, we wrap it into an inner sequence with assign side effect and construct an outer sequence for the whole call and get: `{ locals: temp1, temp2; result: string.Concat(s, { sideEffects: temp1 = c1; result: new ReadOnlySpan<char>(in temp1) }, { sideEffects: temp2 = c2; result: new ReadOnlySpan<char>(in temp2) }) }`
- Similary on the third iteration we unwrap previous result, get `temp1` and `temp2` as previous locals and 3 arguments from that call, wrap the remaining `c3` into a sequence with its own local and construct the final result: `{ locals: temp1, temp2, temp3; result: string.Concat(s, { sideEffects: temp1 = c1; result: new ReadOnlySpan<char>(in temp1) }, { sideEffects: temp2 = c2; result: new ReadOnlySpan<char>(in temp2) }, { sideEffects: temp3 = c3; result: new ReadOnlySpan<char>(in temp3) }) }`

So in the end we can see that:
- When performingthe initial concat of 2 arguments there can be no previous locals
- When "upgrading" concat from 2 to 3 and from 3 to 4 arguments at most 1 additional local can appear

So I modified code according to these observations in the following way:
- For `RewriteStringConcatenationTwoExprs` we no longer pass previous locals builder, assert that it is empty when calling this method and assert that at least 1 of potential locals is `null`
- For `RewriteStringConcatenationThreeExprs` and `RewriteStringConcatenationFourExprs` changed capacity of the locals builder to be `previousLocals.Count + 1` and lowered the assert of `previousLocals.Count` by one on the caller side

All string concatenation tests still pass without hitting these asserts